
## **Gene 760 – Problem Set 4**


The goal of this problem set is to familiarize students with the manipulation and analysis of human genetics data. In this assignment, you will profile genome variation information and attempt to answer biologically relevant questions. 

You’ve been provided with a VCF file (*Trio-NA12891-NA12892-NA12878.vcf.gz*). This file contains SNPs and indels from three individuals. Individuals NA12891, NA12892, and NA12878 form a trio (dad, mom, and child) that you will be analyzing in depth later. NA12878 has been sequenced and analyzed countless times and has been awarded the notorious title of “The Most well studied Genome in the World”.

The variant call format (VCF) is a generic text file format for storing genome variation data such as SNPs, indels, and structural variants, together with rich annotations. It contains meta-information lines, a header line, and then data lines each containing information about a position in the genome. The format can also contain genotype information for multiple samples for each position, which are stored as additional columns. A VCF file is usually stored in a compressed manner and can be indexed for fast data retrieval of variants from a range of positions on the reference genome.



***

The files for PS4 are available in `/gpfs/gibbs/project/gene760/shared/folders/Data/PS4/`, including:

* Trio VCF: (*Trio-NA12891-NA12892-NA12878.vcf.gz*)

* Pre-built gnomAD gnotate file: (*gnomad.genomes.v3.1.sites.chr1-22_XYM.zip*)

* Genome in a Bottle difficult regions: (*GRCh38_alldifficultregions.bed*)

***

You will use the following software:
* BCFtools (https://samtools.github.io/bcftools/bcftools.html)

* Slivar (https://github.com/brentp/slivar)

* BEDtools (https://bedtools.readthedocs.io/en/latest/)

* VEP (https://useast.ensembl.org/Homo_sapiens/Tools/VEP)

 _Note that Slivar is not available as a module. The software binary is available under */gpfs/gibbs/project/gene760/shared/folders/Data/PS4/slivar*_
    
***

Students should submit:
> *<Your_NetID>_PS4_answers.md: Answers to the questions below; your commands should be commented with their (intended) purpose!*

> *<Your_NetID>_mendel.R: R script for Question 6*

> *<Your_NetID>_denovo_variants.R: R script for Question 7*

to `/gpfs/gibbs/project/gene760/shared/Dropbox/PS4` on **Friday, April 5 at 11:59 pm**.

*** 

### **1.** Let’s get some basic statistics of variants in the VCF file. Use BCFtools to answer following questions.
  
#### a. What fraction of variants are SNPs or indels?

```
Replace this line with your answer
```

```
Replace this line with the command(s) you used
```

#### b. Count the number of SNPs for each substitution type. Report it in the form of a table with the six possibilities and their corresponding counts. [Hint: Considering that DNA is double stranded, A -> C is same as T->G and can be written as A/T -> C/G].

```
Replace this line with your answer
```

#### c. Among SNPs, what substitution is the most prevalent? Why is this substitution type the most prevalent?

```
Replace this line with your answer
```


***

### **2.** Since the human genome is diploid, a variant can be found on either (heterozygous) or both (homozygous alternate) chromosomes. Use BCFtools to extract data for the individual NA12878 and perform the following tasks.

#### a. Count the total number of genotype calls that are homozygous reference, heterozygous, and homozygous alternate for that individual.

```
Replace this line with your answer
```

```
Replace this line with the command(s) you used
```


***

### **3.** The allele frequency indicates how often a variant can be observed if we sequence many individuals, and it can be obtained from resources such as gnomAD or the 1000 Genomes Project (1KGP). Use Slivar and the pre-built annotation file (*gnomAD gnotate file*: gnomad.genomes.v3.1.sites.chr1-22_XYM.zip) to annotate the trio VCF file with allele frequencies from gnomAD; store it as Trio-NA12891-NA12892-NA12878.gnomAD_annotated.vcf.gz. 

### The VCF file is already pre-annotated with allele frequencies (INFO field, ID=AF) from 1KGP. ***The allele frequency from 1KGP is to be used for Q3 in case there are troubles with gnomAD annotation***. Write a script (use BCFtools if needed) to perform the following tasks.

#### a. Write the code to annotate the trio VCF file with gnomAD allele frequency using Slivar:

(Note that we expect some warnings : don't panic.)

```
Replace this line with the command(s) you used
```

#### b. Count the fraction of variants that are common (>5%), low allele frequency (1-5%) and rare (<1%) per 1KGP allele frequency:

```
Replace this line with your answer
```

```
Replace this line with the command(s) you used
```

#### c. Count the number of variants that are rare in *at least one* of the five super-populations in 1KGP (AFR, AMR, EAS, EUR, SAS):

```
Replace this line with your answer
```

```
Replace this line with the command(s) you used
```

***

### **4.** The VCF file is annotated with predicted functional consequences (INFO field, ID=CSQ) of alternate alleles (non-reference) by VEP (https://useast.ensembl.org/Homo_sapiens/Tools/VEP). Write a script to answer the following questions (use BCFtools if needed).

#### a. Briefly explain why many variants are annotated with multiple consequences for a single gene:

```
Replace this line with your answer
```

#### b. Count the number of variants in the following categories: synonymous, missense, and probable loss of function (pLOF: stop gained, frameshift, splice_donor/acceptor). Does this count make sense biologically? Why?

(hint: https://useast.ensembl.org/info/genome/variation/prediction/predicted_data.html)
(Note also that it's OK to count a variant which belongs in multiple categories multiple times.)

```
Replace this line with your answer
```

```
Replace this line with the command(s) you used
```

***


### **5.** Mapping short reads is difficult in certain genomic regions, particularly due to repeats. The Genome in a Bottle consortium (GIAB) identified such regions of the human genome (GRCh38_alldifficultregions.bed; BED file format). This file can be used with other benchmarking tools to identify true/false positive variant calls. Use BEDtools to calculate the fraction of variants that overlap with the GIAB difficult regions. Briefly, how do you interpret these results?

(Note: do not apply this filtering to analysis at subsequent steps: )

```
Replace this line with your answer
```

```
Replace this line with the command(s) you used
```

***

### **6.** Write a R script to count the number of variants that clearly violate the rules of Mendelian inheritance, given the trio’s relationships to one another. (See diagram in the HTML version on canvas.)

[![`Mendelian Inheritance`](https://i.imgur.com/lMdkj5b.jpeg)](https://i.imgur.com)

#### a. How many variants violate the rules of Mendelian inheritance?

```
Replace this line with your answer
```

#### b. Describe some potential reasons that could explain these violations:

```
Replace this line with your answer
```

***

### **7.** De novo mutations are those that are present in a child and absent in both parents, potentially occurring as a mutation during gametogenesis during the parent. Write a R script to find _de novo_ variants in NA12878. To do this, you should identify variants that are heterozygous in the child but homozygous reference in both parents. The _de novo_ variants should be written in a new VCF file named Trio-NA12891-NA12892-NA12878.denovo.vcf. Use your preferred method to answer the following questions.

(Note: you may wish to run this on your annotated file to facilitate analysis in 7b.)

#### a. How many _de novo_ variants are observed?

```
Replace this line with your answer
```

```
Replace this line with the command(s) you used
```

#### b. How many _de novo_ variants are present in gnomAD? How many are in the GIAB difficult regions of the genome? Should we be more confident or less confident that these variants (present in gnomAD or overlap with GAB difficult regions) are true _de novo_ variants? Why?

(You can assume that those variants observed in gnomad have passed quality control filters and are compatible with life + not sequencing errors. )

```
Replace this line with your answer
```

```
Replace this line with the command(s) you used
```

***
